### PR TITLE
Bug 4587: socket/memory leak

### PR DESCRIPTION
**Pull request by Opendium**
In situations where the client has made a CONNECT request, if
the ICAP server responds to the REQMOD by returning a new HTTP
response, Squid logs "abandoning" and stops reading from the
socket. Eventually the client drops the connection, which drops
into the CLOSE_WAIT state. Since Squid never services the
socket, it is never closed and stays in the CLOSE_WAIT state
forever.  Associated memory is also not freed, resulting in a
significant memory leak.

Some situations where the ICAP server may want to return a new
HTTP response for a CONNECT are:
 * To forbid the connection by returning a 403 Forbidden.
 * To demand authentication by returning a 407 Proxy 
   Authorisation Required, in situations where the ICAP server,
   rather than Squid, is handling authentication.

This change does 3 things:
 1. If REQMOD responds with an HTTP 200 response, an error is
    logged since this is a bug.
 2. If REQMOD responds with any other HTTP response, set
    readMore so that Squid can process a new request after
    returning the HTTP response.
 3. If the "abandoning" state is reached, close the socket.

Bug #4587 also discusses a situation where the ICAP server
modifies the HTTP request such that it is no longer a CONNECT.
This change does not provide comprehensive support for doing
that, but does at least ensure that the socket and memory are
not leaked.

In the discussion regarding this bug, concerns were raised that
the "abandoning" state may be reached for other unknown cases
where closing a connection would kill a valid transaction.  If
that is a significant concern, that part of this change can be
removed, however, I argue that:
 1. We have been running many systems in the wild since writing
    the original (much less complete) patch attached to the bug
    report over 5 years ago, and have never knowlingly had a
    problem with valid transactions ending up in the
    "abandoning" state.
 2. It is surely better to occasionally kill an extremely rare
    valid transaction than to leak memory and sockets until the
    entire proxy crashes.

I agree that it would be nice to properly characterise what
transactions may end up in the "abandoning" state, but until
that is done I feel that closing those connections is the lesser
of the two evils.

### CODE REVIEW

File: src/client_side_request.cc [hunk 1]
```diff
@@ -2022,6 +2022,23 @@ ClientHttpRequest::handleAdaptedHeader(Http::Message *msg)
...
+        if (request->method == Http::METHOD_CONNECT) {
+            // CONNECT requests that REQMOD produces a reply for need special
+            // handling.
+            if (new_rep->sline.status() == Http::scOkay) {
+                // REQMOD replying with a 200 is a bug, since Squid won't have
+                // connected to the origin server.  The request will hit the "abandoning"
+                // section in ConnStateData::kick().
+                debugs(85, DBG_IMPORTANT, HERE << "REQMOD produced '200 Ok' reply for a CONNECT request");
+            } else {
+                // Returning an error to the client, so turn readMore back on
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/client_side_request.cc:2036]: As you say they should not have any effect. I would like to be sure that Squid is not turning them into a final response or handling them as one inadvertently.

**Review by yadij — APPROVED** [review]

**Review by rousskov — CHANGES_REQUESTED** [review]
Thank you for working on fixing this difficult problem!

The PR description got a bit stale, and the PR code changes will need some low-level polishing to keep the surrounding code in sync and to follow Squid coding style. I can help with those minor fixes, but I would like to resolve a few high-level concerns first.

File: src/client_side_request.cc [hunk 2]
```diff
@@ -2022,6 +2022,25 @@ ClientHttpRequest::handleAdaptedHeader(Http::Message *msg)
     } else if (HttpReply *new_rep = dynamic_cast<HttpReply*>(msg)) {
         debugs(85,3,HERE << "REQMOD reply is HTTP reply");
 
+        if (request->method == Http::METHOD_CONNECT) {
+            // CONNECT requests that REQMOD produces a reply for need special
+            // handling.
+            if (Http::Is2xx(new_rep->sline.status())) {
+                // REQMOD replying with a 2xx is a bug, since Squid won't have
+                // connected to the origin server.
+                debugs(85, DBG_IMPORTANT, HERE << "REQMOD produced a 2xx HTTP response for a CONNECT request");
```
[the remark this answers is not in the document]
Comment (Opendium) [reply src/client_side_request.cc:2031]: Apologies for leaving my PRs alone for a while - I've been stuck with some other work and expect to return to them soon.  However, here are some answers:

### Treating 2xx responses in the same way as any other response:

If we treat a 2xx response in the same way as any other response here, the 2xx response will be forwarded to the client, and then (in the case of keep-alive connections) Squid will wait for a new HTTP request.

This is a broken situation - the client has been told that it now has a CONNECT tunnel to an origin server, whilst Squid is expecting another HTTP request which is unrelated to the CONNECT.  The client is probably trying to access an HTTPS URI, so it will send a TLS Client Hello instead of the HTTP request that Squid was expecting.

I decided that it would be better to produce an error rather than losing synchronisation between Squid and the client.

### What could we do with 2xx responses?

Firstly, without adaptation meddling with the CONNECT, the connection would potentially be handled by ssl bump anyway.  So in order to bump the connection and then have adaptation handle the HTTP requests within it, we wouldn't need to do any adaptation on the CONNECT request itself.

Therefore, my presumption is that the only reason why adaptation might to produce a 2xx HTTP response for a CONNECT is to allow access to the raw data stream, rather than to facilitate any SSL bump related things. (*)

At the point that REQMOD adaptation has generated a 2xx response, Squid has not established a tunnel to anywhere, but we've told the client that one has been established.  So what _could_ we do?

We could establish a tunnel to somewhere.  But there are a couple of problems:
1. Where would we establish the tunnel to? There is no information in the ICAP response indicating where we should connect.
2. If we try to establish a tunnel but it fails, we're in trouble because we already told the client it had succeeded.
Note that instead of generating an HTTP reply as we are doing here, REQMOD can replace the HTTP request with a new CONNECT and allow Squid's normal logic to attempt to establish the tunnel.

Alternatively, we could connect the tunnel to the adaptation service.  Squid doesn't currently have any code to allow an adaptation service to see, process or generate the raw data within the CONNECT tunnel itself.  In the future it would be nice for adaptation to be able to handle raw tunnelled data (both for CONNECT and things like WebSockets), but there is no support for that at the moment, and I expect the ICAP protocol would need to be extended to cope with bidirectional data streams.

So in short, whilst we _could_ make Squid do something useful when ICAP generates a 2xx HTTP response to a CONNECT, it is not at all clear what that useful thing might be.  Whatever useful thing is decided upon, it would certainly require a lot of work to implement.

(*) Alternatively, we could trigger client-first bumping and then insist that all requests within the bumped tunnel result in REQMOD generating an HTTP reply since there we're not connected to any real origin server.  But as mentioned, you can already achieve this in a different way so I don't see a lot of value in it.

[the remark this answers is not in the document]
Comment (rousskov) [reply src/client_side_request.cc:2031]: I agree with most of your points, but they do not imply that an ICAP service satisfying CONNECT with a `200 Connection Established` response is _buggy_. It may not be compatible with the current Squid code, but that is not a Squid bug or a service bug.

Furthermore, I wonder whether there are special use cases that will work here if we allow them. I am thinking about a situation where, for example:

* the client is establishing a CONNECT tunnel to a plain HTTP-speaking[^1] (no TLS) server on port 443, and
* an ICAP service that knows about that client and wants to intercept/satisfy its messages.
* Squid allows but does not SslBump this CONNECT tunnel.

[^1]: Or it could just look like plain HTTP to Squid but in fact it is a different protocol on top of HTTP that the ICAP service is aware of.

AFAICT, at least on the surface of things, Squid can handle that situation without extra code/features -- all the post-CONNECT traffic will be HTTP requests that Squid will parse and forward to the ICAP REQMOD service, and the ICAP service will _satisfy_ them.

If the above hypothetical use case works, the question becomes: Why ban that (and possibly other) use cases? The answer could be that most adaptation services that satisfy CONNECT requests are probably broken/incompatible and, if we allow this use case, we will significantly complicate triage of those cases. Thus, we block all 2xx responses and wait for somebody with genuine need to add a configuration option to allow them. Is that our rationale here?

File: src/client_side_request.cc [hunk 3]
```diff
@@ -2022,6 +2022,25 @@ ClientHttpRequest::handleAdaptedHeader(Http::Message *msg)
...
+                static const auto d = MakeNamedErrorDetail("CLT_REQMOD_2xx_HTTP_RESPONSE");
+                handleAdaptationFailure(d, false);
+                return;
+            } else {
+                // Returning an error to the client, so turn readMore back on
+                // so we can keep the connection alive for the next request.
+                debugs(85,3,HERE << "REQMOD reply rejected a CONNECT");
+                ConnStateData * conn = getConn();
+                if (conn) conn->flags.readMore = 1;
```
[the remark this answers is not in the document]
Comment (Opendium) [reply src/client_side_request.cc:2040]: Yes, this is so that we don't get abandoned.  I think you're probably right that we should call mayUseConnection(false) here.  Oddly, I'm using (almost) this code in the wild on Squid 5 and it is working.

File: src/client_side.cc [hunk 4]
```diff
@@ -988,6 +988,7 @@ ConnStateData::kick()
     } else {
         // XXX: Can this happen? CONNECT tunnels have deferredRequest set.
         debugs(33, DBG_IMPORTANT, MYNAME << "abandoning " << clientConnection);
+        clientConnection->close();
```
[the remark this answers is not in the document]
Comment (Opendium) [reply src/client_side.cc:991]: My PR eliminates the only case I have found where the "abandoning" state is reached, so that part of the PR can be removed if it is of concern.

It was included because hitting the "abandoning" state is:
1. Likely to result in a socket being leaked, which will ultimately crash the server; and
2. Not hit during normal operation, in my experience.

I felt that, although the behaviour isn't characterised well enough to know for sure that legitimate requests will never end up in that state, it is preferable to drop the occasional legitimate request than to crash the server.

[the remark this answers is not in the document]
Comment (rousskov) [reply src/client_side.cc:991]: Please do not misinterpret my question as me fishing for ammunition to undo this change. I was aware of your original motivation here (thank you for documenting it!) and just needed a bit more info to get the remaining code in sync with this change. Thank you for confirming that you no longer see these "abandoning" messages. I am resolving this high-level change request.

File: src/client_side_request.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/client_side_request.cc:2040]: Yeah, Stream::mayUseConnection_ documentation clearly implies that we should make mayUseConnection() false after we decide not to open a tunnel:

```C++
    bool mayUseConnection_; /* This request may use the connection. Don't read anymore requests for now */
```

I am not sure that documentation is 100% accurate, but I see no specific reasons not to trust it.

To reduce chances of this code getting out of sync again, I suggest adding a pair of functions or methods: reserveForTunnel() and cancelTunnelReservation() or perhaps unreserveForTunnel(). The existing "Let tunneling code be fully responsible for CONNECT requests" code will call the former, while your new code (and perhaps other code that decides not to tunnel?) will call the latter. What do you think?


>  Oddly, I'm using (almost) this code in the wild on Squid 5 and it is working.

Well, your "wild" may not be wild enough (or you just do not see some of the subtle failures). These flags become especially important when clients pipeline requests or do other relatively unusual/rare things. In the majority of use cases that I see in the code, mayUseConnection_ is checked before this PR code runs. ConnStateData::terminateAll() comes later, but it only runs in special cases.

File: src/client_side_request.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (Opendium) [reply src/client_side_request.cc:2031]: > I agree with most of your points, but they do not imply that an ICAP service satisfying CONNECT with a 200 Connection Established response is buggy. It may not be compatible with the current Squid code, but that is not a Squid bug or a service bug.

There is no specification that I'm aware of that specifies what should actually happen where an ICAP service produces a "200 Connection Established" response (in terms of handling the connection after the response it returned to the client).  So any ICAP service that does this is now outside the realms of specified behaviour - you have to know how the specific ICAP client will handle that response.  Maybe "buggy" is too strong a term, but it is undefined behaviour (which may be defined in the future, and the eventual definition may not match what we implement :).

> AFAICT, at least on the surface of things, Squid can handle that situation without extra code/features -- all the post-CONNECT traffic will be HTTP requests that Squid will parse and forward to the ICAP REQMOD service, and the ICAP service will satisfy them.

I don't believe that would work, because the client would be making origin server HTTP requests instead of proxy HTTP requests.  i.e. the requests would look like "GET /" instead of "GET http://example.com/".

We could certainly make Squid handle these requests (in particular, it would have to make an assumption about the URI scheme), but I don't believe it will work out of the box.

I agree that producing an error helps triage ICAP services which are producing a response which results in undefined behaviour.  I don't believe that allowing 2xx replies is of much use without further modifications to squid, but I am happy to remove the ban on 2xx replies if that's the preferred way forward.

File: src/client_side_request.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (Opendium) [reply src/client_side_request.cc:2040]: > To reduce chances of this code getting out of sync again, I suggest adding a pair of functions or methods: reserveForTunnel() and cancelTunnelReservation() or perhaps unreserveForTunnel(). The existing "Let tunneling code be fully responsible for CONNECT requests" code will call the former, while your new code (and perhaps other code that decides not to tunnel?) will call the latter. What do you think?

Sounds like a good plan.

> Well, your "wild" may not be wild enough (or you just do not see some of the subtle failures). These flags become especially important when clients pipeline requests or do other relatively unusual/rare things.

That's entirely possible.  In my case, my ICAP REQMOD service is responsible for deciding which requests to authenticate, and it returns a suitable "407" HTTP response(*).  Clients will regularly reuse the connection to send the authenticated request, but you're right that we're probably not seeing pipelined requests.

(* previously we were using an external ACL to talk to the ICAP server and control authentication, but eventually decided that we could dispense with it by just having the ICAP server produce the 407 response itself.  We've been using the new method for a few months now and it seems to work well.)

File: src/client_side_request.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/client_side_request.cc:2031]: Thank you for a thoughtful response. I have doubts about some of the assertions in your statements, but I am OK with banning these 2xx responses (at least for now) as long as we do not label them as bugs.

I suggest giving the level-1 message an ID, so that admins can control the message frequency via cache_log_message -- the affected ICAP service is likely to emit many such messages.

Do you want to take a stub at the corresponding changes?

* If yes, please also drop the deprecated `HERE`. I would also use a "REQMOD satisfied CONNECT with an HTTP 2xx response" wording to emphasize that we are dealing with request satisfaction here. Technically, a REQMOD service can modify origin server responses as well (but Squid does not support post-cache REQMOD yet).
* If no, I can help with these adjustments.

[the remark this answers is not in the document]
Comment (Opendium) [reply src/client_side_request.cc:2031]: I'm happy to implement the changes.  It won't be this week though I'm afraid.

[the remark this answers is not in the document]
Comment (rousskov) [reply src/client_side_request.cc:2031]: Thank you. No rush at all!
